### ISSUE TITLE
[QUESTION] - Unable to Migrate the SBQQ__Quote__c object data 

### ISSUE THREAD

**Issue by kvenkatasivareddy** [forcedotcom/SFDX-Data-Move-Utility#623]
Hi team, 
Firstly i want to thank you guys for building such an amazing tool to migrate the data. 

My issue is related to https://github.com/forcedotcom/SFDX-Data-Move-Utility/issues/315.

I want to migrate the data of SBQQ__Quote__c. 
[With SBQQ__PricebookId__c Field.zip](https://github.com/forcedotcom/SFDX-Data-Move-Utility/files/12247903/With.SBQQ__PricebookId__c.Field.zip)

I tried excluding the SBQQ__PricebookId__c field from SBQQ__Quote__c SOQL query it started working. 
[Without SBQQ__PricebookId__c Field.zip](https://github.com/forcedotcom/SFDX-Data-Move-Utility/files/12247905/Without.SBQQ__PricebookId__c.Field.zip)

I want to migrate the SBQQ__PricebookId__c field along with SBQQ__Quote__c object. So as per your comments in the https://github.com/forcedotcom/SFDX-Data-Move-Utility/issues/315#issuecomment-849114243 i tried adding field mapping for SBQQ__PricebookId__c. Again migration is getting failed with following configuration
[With FieldMapping for SBQQ__Quote__c Failed.zip](https://github.com/forcedotcom/SFDX-Data-Move-Utility/files/12248067/With.FieldMapping.for.SBQQ__Quote__c.Failed.zip)

Can you please help to migrate the SBQQ__PricebookId__c field along with SBQQ__Quote__c?

Also i have below question in my mind.
1) can't we map the single column to multiple columns within a same object using fieldMapping?
2) After excluding the SBQQ_PricebookId__c field from SBQQ__Quote__c object, In the autogenerated csv files under target folder I see the SBQQ__PriceBook__c field has the data. But the same data not getting populated in salesforce UI or Developer Console. Is this is because of excluding the SBQQ__PricebookId__c or it is a bug?
![image](https://github.com/forcedotcom/SFDX-Data-Move-Utility/assets/113102274/0d273b11-e29b-4ffc-aa8a-af4b2f2f611e)
[Opportunity_insert_target.csv](https://github.com/forcedotcom/SFDX-Data-Move-Utility/files/12250011/Opportunity_insert_target.csv)
[SBQQ__Quote__c_insert_target.csv](https://github.com/forcedotcom/SFDX-Data-Move-Utility/files/12250012/SBQQ__Quote__c_insert_target.csv)
[SBQQ__Quote__c_update_target.csv](https://github.com/forcedotcom/SFDX-Data-Move-Utility/files/12250013/SBQQ__Quote__c_update_target.csv)
[2023-08-03__16_06_25.log](https://github.com/forcedotcom/SFDX-Data-Move-Utility/files/12250018/2023-08-03__16_06_25.log)

Thanks in advance:)

**Comment by kvenkatasivareddy** [comment]
Thank you for responding. Waiting for your response!!

**Comment by kvenkatasivareddy** [comment]
Hi @hknokh, 
Any update?

**Comment by hknokh2** [comment]
Hello
For now I haven't any update for you.
I will let you know if any.
Regards

**Comment by kvenkatasivareddy** [comment]
Yes, you are correct.
I want sfdmu to update the value in SBQQ__PricebookId__c to the actual Id of the pricebook in the current environment.

**Comment by hknokh2** [comment]
Hello @mattgrimwade,

Thank you for the explanation. You're spot on!

@kvenkatasivareddy,

Here's some clarity on field mapping:

- **Field Mapping Explained:** When you set up field mapping, the fields and objects are mapped **between the source and target orgs.** They're not mapped between target and target. In simpler terms, field mapping signifies: "which target object and field will be updated instead of the corresponding source object and field."

Using your provided example:
```json
"fieldMapping": [
  {
    "targetObject": "SBQQ__Quote__c"
  },
  {
    "sourceField": "Id",
    "targetField": "SBQQ__PricebookId__c"
  }
],
```
The actual mapping translates to:

- Pricebook2.IsActive => SBQQ__Quote__c.IsActive
- Pricebook2.Name => SBQQ__Quote__c.Name
- Pricebook2.Description => SBQQ__Quote__c.Description
- Pricebook2.Id => SBQQ__Quote__c.SBQQ__PricebookId__c

- **Automatic Field Mapping:** When mapping to a different sObject, **all fields of the source sObject are automatically mapped to their corresponding fields in the target object,** barring any that are manually mapped using the "sourceField => targetField" pairs.

- **Regarding post-mapping:** There isn't an out-of-the-box (OOTB) method to set up post-mapping from target to target after the target has been updated. This isn't the primary purpose of field mapping. While thinking of a workaround, you might want to consider creating a custom add-on for the SFDMU. On my end, I've added the idea of target-to-target mapping to my development backlog, but I can't guarantee a swift implementation.

Best regards

**Comment by kvenkatasivareddy** [comment]
Thanks for considering this as feature request @hknokh. It can be helpful if you implement this soon.

Waiting for this feature.

Thanks in advance:)

**Comment by hknokh** [comment]
Hello,
I hope the issue has been resolved for you. I'm closing it for now. Please let me know if you have any further questions.
Regards.

**Comment by R0Wi** [comment]
@kvenkatasivareddy I think you're right and I also observed that setting the Pricebook reference on `SBQQ__Quote__c` can only be done by uploading it in the `SBQQ__Quote__c.SBQQ__PricebookId__c` textfield rather than using the lookup `SBQQ__Quote__c.SBQQ__PriceBook__c` (which I think SFDMU does automatically). So the solution I found for taking the source `SBQQ__PriceBook__c` and writing it into the target `SBQQ__PricebookId__c` was something like this:

```json
{
    "objectSets" : [
        {
            "objects": [
                {
                    "query": "SELECT IsActive,Name,Description FROM Pricebook2",
                    "operation": "Upsert",
                    "externalId": "Name;Description",
                    "master": false
                },
                {
                    "query": "SELECT SBQQ__ConsumptionRateOverride__c,SBQQ__LineItemsGrouped__c,SBQQ__LineItemsPrinted__c,SBQQ__OrderByQuoteLineGroup__c,SBQQ__Ordered__c,SBQQ__Primary__c,SBQQ__Unopened__c,SBQQ__WatermarkShown__c,SBQQ__Account__c,SBQQ__BillingCity__c,SBQQ__BillingCountry__c,SBQQ__BillingFrequency__c,SBQQ__BillingName__c,SBQQ__BillingPostalCode__c,SBQQ__BillingState__c,SBQQ__BillingStreet__c,SBQQ__ContractingMethod__c,SBQQ__CustomerDiscount__c,SBQQ__DefaultTemplate__c,SBQQ__DeliveryMethod__c,SBQQ__DistributorDiscount__c,SBQQ__Distributor__c,SBQQ__DocumentStatus__c,SBQQ__EmailTemplateId__c,SBQQ__EndDate__c,SBQQ__ExpirationDate__c,SBQQ__FirstSegmentTermEndDate__c,SBQQ__GenerateContractedPrice__c,SBQQ__Key__c,SBQQ__LastCalculatedOn__c,SBQQ__LastSavedOn__c,SBQQ__MarkupRate__c,SBQQ__MasterContract__c,SBQQ__MasterEvergreenContract__c,SBQQ__Opportunity2__c,SBQQ__OrderBy__c,SBQQ__OrderGroupID__c,SBQQ__OriginalQuote__c,SBQQ__PaperSize__c,SBQQ__PartnerDiscount__c,SBQQ__Partner__c,SBQQ__PaymentTerms__c,SBQQ__PriceBook__c,SBQQ__PricebookId__c,SBQQ__PrimaryContact__c,SBQQ__ProrationDayOfMonth__c,SBQQ__QuoteLanguage__c,SBQQ__QuoteProcessId__c,SBQQ__QuoteTemplateId__c,SBQQ__RenewalTerm__c,SBQQ__RenewalUpliftRate__c,SBQQ__SalesRep__c,SBQQ__ShippingCity__c,SBQQ__ShippingCountry__c,SBQQ__ShippingName__c,SBQQ__ShippingPostalCode__c,SBQQ__ShippingState__c,SBQQ__ShippingStreet__c,SBQQ__Source__c,SBQQ__StartDate__c,SBQQ__Status__c,SBQQ__SubscriptionTerm__c,SBQQ__TargetCustomerAmount__c,SBQQ__Type__c FROM SBQQ__Quote__c LIMIT 50",
                    "operation": "Upsert",
                    "externalId": "SBQQ__BillingCity__c;SBQQ__BillingCountry__c;SBQQ__BillingName__c;SBQQ__BillingPostalCode__c;SBQQ__BillingState__c;SBQQ__CustomerDiscount__c;SBQQ__DistributorDiscount__c;SBQQ__EmailTemplateId__c;SBQQ__Key__c;SBQQ__MarkupRate__c;SBQQ__OrderGroupID__c;SBQQ__PartnerDiscount__c;SBQQ__PricebookId__c;SBQQ__QuoteProcessId__c;SBQQ__QuoteTemplateId__c;SBQQ__RenewalUpliftRate__c;SBQQ__ShippingCity__c;SBQQ__ShippingCountry__c;SBQQ__ShippingName__c;SBQQ__ShippingPostalCode__c;SBQQ__ShippingState__c",
                    "beforeUpdateAddons" : [
                        {                        
                            "module" : "core:RecordsTransform",
                            "args": {
                                "fields": [
                                    {
                                        "alias": "pricebookId",
                                        "sourceObject": "SBQQ__Quote__c",
                                        "sourceField": "SBQQ__PriceBook__c"
                                    }
                                ],
                                "transformations": [
                                    {
                                        "targetObject": "SBQQ__Quote__c",
                                        "targetField": "SBQQ__PricebookId__c",
                                        "formula": "formula.pricebookId"
                                    },
                                    {
                                        "targetObject": "SBQQ__Quote__c",
                                        "targetField": "SBQQ__PriceBook__c",
                                        "formula": "''"
                                    }
                                ]
                            }
                        }
                    ]
                }
            ]
        }
    ]
}
```

You can see I'm using the [record transformation](https://help.sfdmu.com/examples/configuration/record-transformation) module to set the `SBQQ__PricebookId__c` and to clear the `SBQQ__PriceBook__c` before uploading to Salesforce.

Maybe this approach will unblock you and you won't have to wait for any additional SFDMU feature, good luck 😄

**Comment by kvenkatasivareddy** [comment]
thank you @R0Wi
